# Capstone Project - Connections between Lifestyle & Disease

# Data wrangling & Exploratory Data Analysis (EDA)

In [2]:
# Import packages.

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
# Import data.

df = pd.read_csv("AG_copy_1.csv")

/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (34,197) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [4]:
# Show top of data frame.

df.head(5)

#SampleID BarcodeSequence LinkerPrimerSequence CARBOHYDRATE_PER  \
0  11224.12109         no_data              no_data               55   
1  5034.107635         no_data              no_data               56   
2  11351.12107         no_data              no_data               16   
3  2013.113155         no_data              no_data          no_data   
4  4731.107641         no_data              no_data               61   

  NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS ASSIGNED_FROM_GEO RUN_DATE  \
0               no_data                    no_data                 n  no_data   
1               no_data                    no_data                 n  no_data   
2               no_data                    no_data                 n  no_data   
3               no_data                    no_data                 n  no_data   
4                   yes                    no_data                 n  no_data   

  PET_LOCATIONS AGE  ... LIBRARY_CONSTRUCTION_PROTOCOL PKU DECEASED_PARENT  \
0    Housing...  51  ...                       no_data  no         no_data   
1       no_data  35  ...                       no_data  no         no_data   
2    Housing...  56  ...                       no_data  no         no_data   
3    Housing...  37  ...                       no_data  no         no_data   
4       no_data  72  ...                       no_data  no         no_data   

   SEQ_RUN  CAT LATITUDE FOODALLERGIES_PEANUTS CHICKENPOX SEASONAL_ALLERGIES  \
0  no_data  yes     39.7               no_data        yes                 no   
1  no_data  yes     32.7               no_data        yes                 no   
2  no_data   no       40               no_data        yes                 no   
3  no_data   no     40.7               no_data        yes                 no   
4  no_data   no     43.1               no_data        yes                yes   

                         Description  
0  American Gut Project Stool sample  
1  American Gut Project Stool sample  
2  American Gut Project Stool sample  
3  American Gut Project Stool sample  
4  American Gut Project Stool sample  

[5 rows x 202 columns]

In [5]:
# Show object type.

type(df)

pandas.core.frame.DataFrame

In [6]:
# Show shape of dataframe. 

df.shape

(4815, 202)

The data set has 4815 rows and 202 features. 

---

In [7]:
# Show the data types of all columns.

df.dtypes

#SampleID                object
BarcodeSequence          object
LinkerPrimerSequence     object
CARBOHYDRATE_PER         object
NONFOODALLERGIES_DRUG    object
                          ...  
LATITUDE                 object
FOODALLERGIES_PEANUTS    object
CHICKENPOX               object
SEASONAL_ALLERGIES       object
Description              object
Length: 202, dtype: object

In [8]:
# Show all the non-numerical/categorical columns.

df.dtypes[df.dtypes != 'int64'][df.dtypes != 'float64']

# https://stackoverflow.com/questions/40353079/pandas-how-to-check-dtype-for-all-columns-in-a-dataframe

#SampleID                object
BarcodeSequence          object
LinkerPrimerSequence     object
CARBOHYDRATE_PER         object
NONFOODALLERGIES_DRUG    object
                          ...  
LATITUDE                 object
FOODALLERGIES_PEANUTS    object
CHICKENPOX               object
SEASONAL_ALLERGIES       object
Description              object
Length: 200, dtype: object

200 features are categorical (object).

In [9]:
# Show all the numerical columns.

df.dtypes[df.dtypes != 'object']

ALTITUDE    int64
DEPTH       int64
dtype: object

2 features are numerical (int64). 

In [10]:
# Show info about numerical columns.

df.describe()

ALTITUDE   DEPTH
count  4815.000000  4815.0
mean      0.000208     0.0
std       0.014411     0.0
min       0.000000     0.0
25%       0.000000     0.0
50%       0.000000     0.0
75%       0.000000     0.0
max       1.000000     0.0

In [11]:
# Show vales of numerical columns. 

print(df[['ALTITUDE', 'DEPTH']])

      ALTITUDE  DEPTH
0            0      0
1            0      0
2            0      0
3            0      0
4            0      0
...        ...    ...
4810         0      0
4811         0      0
4812         0      0
4813         0      0
4814         0      0

[4815 rows x 2 columns]


Initial inspection shows that the altitude and depth columns might be (mostly or only) zeros.

In [12]:
# Count amount of distinct values in a column.

df['ALTITUDE'].nunique()

# https://stackoverflow.com/questions/45759966/counting-unique-values-in-a-column-in-pandas-dataframe-like-in-qlik

2

In [13]:
# Show frequency of distinct values in a column.

df['ALTITUDE'].value_counts()

# https://stackoverflow.com/questions/22391433/count-the-frequency-that-a-value-occurs-in-a-dataframe-column

0    4814
1       1
Name: ALTITUDE, dtype: int64

In [14]:
# Count amount of distinct values in a column.

df['DEPTH'].nunique()

1

In [15]:
# Show frequency of distinct values in a column.

df['DEPTH'].value_counts()

0    4815
Name: DEPTH, dtype: int64

Further inspection shows that the altitude and depth columns are (mostly or only) zeros. Therefore these columns will be deleted.

---

In [16]:
# Count only non-null values, use 'count'.
# Count total values including null values, use 'size'.
# Count distinct values, use 'nunique'.


df_agg = df.agg(['count', 'size', 'nunique'])
df_agg

# https://stackoverflow.com/questions/45759966/counting-unique-values-in-a-column-in-pandas-dataframe-like-in-qlik

#SampleID  BarcodeSequence  LinkerPrimerSequence  CARBOHYDRATE_PER  \
count         4815             4815                  4815              4815   
size          4815             4815                  4815              4815   
nunique       4815              257                     2               159   

         NONFOODALLERGIES_DRUG  NONFOODALLERGIES_BEESTINGS  ASSIGNED_FROM_GEO  \
count                     4815                        4815               4815   
size                      4815                        4815               4815   
nunique                      3                           3                  2   

         RUN_DATE  PET_LOCATIONS   AGE  ...  LIBRARY_CONSTRUCTION_PROTOCOL  \
count        4815           4815  4815  ...                           4815   
size         4815           4815  4815  ...                           4815   
nunique         3             48    90  ...                              2   

          PKU  DECEASED_PARENT  SEQ_RUN   CAT  LATITUDE  \
count    4815             4815     4815  4815      4815   
size     4815             4815     4815  4815      4815   
nunique     4                3        5     4       430   

         FOODALLERGIES_PEANUTS  CHICKENPOX  SEASONAL_ALLERGIES  Description  
count                     4815        4815                4815         4815  
size                      4815        4815                4815         4815  
nunique                      3           4                   4           39  

[3 rows x 202 columns]

In [17]:
# Show the all rows of all features. 

with pd.option_context('display.max_columns', None):
    print (df_agg)

         #SampleID  BarcodeSequence  LinkerPrimerSequence  CARBOHYDRATE_PER  \
count         4815             4815                  4815              4815   
size          4815             4815                  4815              4815   
nunique       4815              257                     2               159   

         NONFOODALLERGIES_DRUG  NONFOODALLERGIES_BEESTINGS  ASSIGNED_FROM_GEO  \
count                     4815                        4815               4815   
size                      4815                        4815               4815   
nunique                      3                           3                  2   

         RUN_DATE  PET_LOCATIONS   AGE  TOT_MASS  GENERAL_MEDS  SAMPLE_ID  \
count        4815           4815  4815      4815          4815       4815   
size         4815           4815  4815      4815          4815       4815   
nunique         3             48    90       132             3       3737   

         ORIG_SAMPLE_NAME  QUINOLINE  BODY_SITE  

For the columns, there is a large range for the number of unique values 'nunique'.

---

In [18]:
# Ask if there are any Nan in the data frame.

df.isnull().values.any()

# https://datatofish.com/check-nan-pandas-dataframe/

True

There are NaN values in the data frame.

In [19]:
# Ask how many Nan there are in the data frame.

df.isnull().sum().sum()

# https://datatofish.com/check-nan-pandas-dataframe/

10

There are 10 rows that contain NaN values.

In [20]:
# Show columns with NaN.

df.loc[:, df.isna().any()]

# https://stackoverflow.com/questions/36226083/how-to-find-which-columns-contain-any-nan-value-in-pandas-dataframe

FOODALLERGIES_OTHER_TEXT ANTIBIOTIC_CONDITION    STATE  \
0                         no_data              no_data       CO   
1                         no_data              no_data       CA   
2     Cats, dogs, certain grasses              no_data       CO   
3                         no_data              no_data       NY   
4                            dust      dental cleaning       NY   
...                           ...                  ...      ...   
4810                      unknown                  UTI       IL   
4811                      unknown      Streptococcus A  unknown   
4812                      no_data              no_data       WA   
4813                      no_data              no_data       IL   
4814                      no_data              no_data   BANKOK   

                    PRIMARY_VEGETABLE  \
0                             spinach   
1     tomatoes (I eat a lot of salsa)   
2                        leafy greens   
3                              Potato   
4                            broccoli   
...                               ...   
4810                  Romaine Lettuce   
4811              tomatoes, cuncumber   
4812                           potato   
4813                     Leafy Greens   
4814                Leaves and fruits   

                                      DIET_RESTRICTIONS  
0                                               no_data  
1                                               no_data  
2                                            Grain free  
3                                               no_data  
4                                               no_data  
...                                                 ...  
4810                                            unknown  
4811                                            unknown  
4812  Ray Peat diet. no grains, beans, minimal veget...  
4813                                            no_data  
4814                                            no_data  

[4815 rows x 5 columns]

Above are the 5 columns that contain NaN values.

In [21]:
# Show rows with NaN.

is_NaN = df.isnull()
row_has_NaN = is_NaN.any(axis=1)
rows_with_NaN = df[row_has_NaN]

print(rows_with_NaN)

# https://www.kite.com/python/answers/how-to-find-rows-with-nan-values-in-a-pandas-dataframe-in-python

               #SampleID BarcodeSequence LinkerPrimerSequence  \
190          13018.12575         no_data              no_data   
208   000014794.fixed934         no_data              no_data   
522   000014791.fixed933         no_data              no_data   
1369  000014793.fixed875         no_data              no_data   
1980          2578.10762         no_data              no_data   
2500         14493.12597         no_data              no_data   
3051         14483.12596         no_data              no_data   
3306  000014792.fixed931         no_data              no_data   
3720         2321.121307         no_data              no_data   
3997         3147.113054         no_data              no_data   

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \
190                69               no_data                    no_data   
208                20               unknown                    unknown   
522                20                   yes                   

Above are the 10 rows that contain NaN values.

First some columns will be deleted. Then the 10 rows with NaN values will be deleted since otherwise, in the inverse order, an error message appears.  

---

Check the value range of a few specially important columns. Here I am choosing columns relevant for diabetes.

In [22]:
# Show unique values in a column.
# Age range.

AGE_unique = np.sort(df.AGE.unique())
AGE_unique

# https://stackoverflow.com/questions/32072076/find-the-unique-values-in-a-column-and-then-sort-them/32072252

array(['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18',
       '19', '2', '20', '21', '22', '23', '24', '25', '26', '27', '28',
       '29', '3', '30', '31', '32', '33', '34', '35', '36', '37', '38',
       '39', '4', '40', '41', '42', '43', '44', '45', '46', '47', '48',
       '49', '5', '50', '51', '52', '53', '54', '55', '56', '57', '58',
       '59', '6', '60', '61', '62', '63', '64', '65', '66', '67', '68',
       '69', '7', '70', '71', '72', '73', '74', '75', '76', '77', '78',
       '79', '8', '80', '81', '82', '83', '84', '88', '9', '91', '93',
       'no_data', 'unknown'], dtype=object)

In [23]:
DIABETES_unique = np.sort(df.DIABETES.unique())
DIABETES_unique

array(['I do not have diabetes', 'Type I', 'Type II', 'no_data',
       'unknown'], dtype=object)

In [24]:
CARBOHYDRATE_PER_unique = np.sort(df.CARBOHYDRATE_PER.unique())
CARBOHYDRATE_PER_unique

array(['0', '1', '10', '11', '11.4', '12', '13', '13.5', '14', '15', '16',
       '17', '17.3', '17.6', '18', '19', '19.2', '19.9', '2', '2.9', '20',
       '20.1', '21', '21.3', '22', '22.2', '23', '24', '25', '25.4', '26',
       '26.8', '27', '27.8', '28', '28.2', '28.3', '29', '29.2', '3',
       '30', '31', '31.4', '31.5', '31.7', '32', '33', '34', '34.2', '35',
       '35.01', '35.6', '36', '36.2', '36.3', '36.6', '36.61', '37', '38',
       '38.1', '38.7', '38.81', '39', '39.1', '39.4', '39.6', '39.8', '4',
       '4.2', '40', '40.2', '41', '41.6', '41.8', '42', '42.1', '42.2',
       '42.3', '42.6', '43', '43.3', '43.4', '43.7', '44', '44.3', '44.5',
       '44.7', '45', '45.4', '45.5', '46', '46.4', '46.5', '46.7', '47',
       '47.1', '47.2', '47.5', '47.6', '47.7', '48', '48.3', '48.5',
       '48.7', '49', '49.6', '49.8', '5', '50', '50.2', '50.8', '51',
       '51.7', '52', '53', '53.5', '54', '55', '55.5', '56', '56.1', '57',
       '57.7', '58', '59', '59.7', '6', '60', 

In [25]:
FAT_PER_unique = np.sort(df.FAT_PER.unique())
FAT_PER_unique

array(['0', '10', '11', '11.7', '12', '13', '13.9', '14', '15', '16',
       '17', '17.7', '18', '19', '19.01', '2', '20', '21', '22', '22.1',
       '23', '24', '25', '25.3', '25.5', '26', '26.8', '26.9', '27', '28',
       '28.3', '28.6', '29', '29.1', '29.4', '30', '30.5', '30.7', '30.9',
       '31', '31.2', '31.4', '31.6', '31.9', '32', '32.4', '32.7', '32.8',
       '33', '33.7', '33.8', '33.9', '34', '34.2', '34.5', '34.6', '35',
       '35.1', '35.4', '35.6', '36', '36.1', '36.2', '36.3', '36.5',
       '36.6', '37', '37.4', '37.5', '37.9', '38', '38.1', '38.3', '38.5',
       '38.8', '39', '39.1', '39.3', '39.5', '39.9', '4', '40', '40.3',
       '40.5', '40.94', '41', '41.2', '41.3', '41.4', '41.7', '42',
       '42.4', '42.6', '43', '43.2', '43.3', '44', '44.1', '44.8', '44.9',
       '45', '45.12', '45.3', '46', '46.58', '46.7', '47', '47.1', '47.2',
       '47.9', '48', '48.7', '49', '5', '50', '51', '51.3', '52', '53',
       '53.5', '54', '54.7', '55', '55.3', '55.8', '5

---

In [26]:
# Look for rows, that contain cells labelled as 'no_data'.

df[(df == 'no_data').any(axis=1)]


/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


#SampleID BarcodeSequence LinkerPrimerSequence  \
0             11224.12109         no_data              no_data   
1             5034.107635         no_data              no_data   
2             11351.12107         no_data              no_data   
3             2013.113155         no_data              no_data   
4             4731.107641         no_data              no_data   
...                   ...             ...                  ...   
4810   000013296.fixed213         no_data              no_data   
4811  000007822.fixed1101         no_data              no_data   
4812          9074.121304         no_data              no_data   
4813          2418.128313         no_data              no_data   
4814          7179.126173         no_data              no_data   

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \
0                  55               no_data                    no_data   
1                  56               no_data                    no_data   
2                  16               no_data                    no_data   
3             no_data               no_data                    no_data   
4                  61                   yes                    no_data   
...               ...                   ...                        ...   
4810               50               unknown                    unknown   
4811               17               unknown                    unknown   
4812               34               no_data                    no_data   
4813               30               no_data                    no_data   
4814               60               no_data                    no_data   

     ASSIGNED_FROM_GEO RUN_DATE PET_LOCATIONS AGE  ...  \
0                    n  no_data    Housing...  51  ...   
1                    n  no_data       no_data  35  ...   
2                    n  no_data    Housing...  56  ...   
3                    n  no_data    Housing...  37  ...   
4                    n  no_data       no_data  72  ...   
...                ...      ...           ...  ..  ...   
4810                 y  no_data    Housing...  71  ...   
4811                 y  no_data    Housing...   9  ...   
4812                 n  no_data    Housing...  29  ...   
4813                 y  no_data       no_data  46  ...   
4814                 y  no_data    Housing...   1  ...   

     LIBRARY_CONSTRUCTION_PROTOCOL      PKU DECEASED_PARENT  \
0                          no_data       no         no_data   
1                          no_data       no         no_data   
2                          no_data       no         no_data   
3                          no_data       no         no_data   
4                          no_data       no         no_data   
...                            ...      ...             ...   
4810                       no_data       no         unknown   
4811                       no_data  unknown         unknown   
4812                       no_data       no         no_data   
4813                       no_data       no         no_data   
4814                       no_data       no         no_data   

                              SEQ_RUN  CAT LATITUDE FOODALLERGIES_PEANUTS  \
0                             no_data  yes     39.7               no_data   
1                             no_data  yes     32.7               no_data   
2                             no_data   no       40               no_data   
3                             no_data   no     40.7               no_data   
4                             no_data   no     43.1               no_data   
...                               ...  ...      ...                   ...   
4810                          no_data  yes       40               unknown   
4811                          no_data   no     50.8               unknown   
4812                          no_data  yes     47.7               no_data   
4813  Knight_American_Gut_Freeze_Thaw  yes       42               no_data   
4814                          no_

In [27]:
df[(df == 'no_data').any(axis=1)].shape

/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


(4815, 202)

In [28]:
# Look for rows, that contain cells labelled as 'No_data'.

df[(df == 'No_data').any(axis=1)]

Empty DataFrame
Columns: [#SampleID, BarcodeSequence, LinkerPrimerSequence, CARBOHYDRATE_PER, NONFOODALLERGIES_DRUG, NONFOODALLERGIES_BEESTINGS, ASSIGNED_FROM_GEO, RUN_DATE, PET_LOCATIONS, AGE, TOT_MASS, GENERAL_MEDS, SAMPLE_ID, ORIG_SAMPLE_NAME, QUINOLINE, BODY_SITE, MIGRAINE_FACTOR_2, MIGRAINE_FACTOR_3, TONSILS_REMOVED, MIGRAINE_FACTOR_1, ELEVATION, FLOSSING_FREQUENCY, PROTEIN_PER, CSECTION, PCR_PRIMERS, COUNTRY_OF_BIRTH, MACRONUTRIENT_PCT_TOTAL, COSMETICS_FREQUENCY, MASTERMIX_LOT, HOST_SUBJECT_ID, ENA-CHECKLIST, GLUTEN, SAMPLE_NAME, TM1000_8_TOOL, LONGITUDE, DOG, MIGRAINE_AGGRAVATION, ORIG_NAME, FRAT, DIABETES_MEDICATION, SAMPLE_PLATE, PENICILLIN, FOODALLERGIES_OTHER_TEXT, KEY_SEQ, BODY_PRODUCT, REGION, SAMPLE_TYPE, MIGRAINE_FREQUENCY, MIGRAINE_RELATIVES, EXPERIMENT_TITLE, PLANT_PER, EXTRACTIONKIT_LOT, SAMPLE, MIGRAINE_PHOTOPHOBIA, SITE_SAMPLED, DRINKING_WATER_SOURCE, CENTER_NAME, COUNTRY, ANTIBIOTIC_SELECT, MIGRAINE_NAUSEA, COMMON_NAME, FOODALLERGIES_TREENUTS, MIGRAINE_AURA, SEQUENCING_METH, FLU_VACCINE_DATE, PREGNANT, MIGRAINE, BMI, ENV_BIOME, WELL_ID, TEETHBRUSHING_FREQUENCY, MIGRAINE_MEDICATIONS, CONTRACEPTIVE, SAMPLE_CENTER, PHYSICAL_LOCATION, LIVINGWITH, LINKER, RELATIONS, DIABETES_MEDICATIONS, SUPPLEMENTS, AGE_UNIT, TANNING_SPRAYS, EXERCISE_FREQUENCY, HOST_COMMON_NAME, SULFA_DRUG, NONFOODALLERGIES_NO, ENV_MATTER, RACE_OTHER, SKIN_CONDITION, ALTITUDE, NONFOODALLERGIES_DANDER, CONDITIONS_MEDICATION, LAST_TRAVEL, TARGET_SUBFRAGMENT, TITLE, HEIGHT_IN, ANTIBIOTIC_CONDITION, PET_CONTACT, MIGRAINE_PAIN, ROOMMATES, ...]
Index: []

[0 rows x 202 columns]

In [29]:
# Look for rows, that contain cells labelled as 'unknown'.

df[(df == 'unknown').any(axis=1)]

/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


#SampleID BarcodeSequence LinkerPrimerSequence  \
1             5034.107635         no_data              no_data   
4             4731.107641         no_data              no_data   
7      000015569.fixed647         no_data              no_data   
9      000011031.fixed716         no_data              no_data   
10     000016722.fixed480    TTCCTGTTAACC  GTGCCAGCMGCCGCGGTAA   
...                   ...             ...                  ...   
4805   000016360.fixed111         no_data              no_data   
4807          3093.107579         no_data              no_data   
4809   000015346.fixed136         no_data              no_data   
4810   000013296.fixed213         no_data              no_data   
4811  000007822.fixed1101         no_data              no_data   

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \
1                  56               no_data                    no_data   
4                  61                   yes                    no_data   
7                  37                   yes                    unknown   
9                  50                   yes                    unknown   
10                 52               unknown                    unknown   
...               ...                   ...                        ...   
4805               19                   yes                    unknown   
4807          no_data               no_data                    no_data   
4809               65               unknown                    unknown   
4810               50               unknown                    unknown   
4811               17               unknown                    unknown   

     ASSIGNED_FROM_GEO RUN_DATE PET_LOCATIONS AGE  ...  \
1                    n  no_data       no_data  35  ...   
4                    n  no_data       no_data  72  ...   
7                    y  no_data       Outdoor  65  ...   
9                    y  no_data    Housing...  54  ...   
10                   y  8/21/14    Housing...  32  ...   
...                ...      ...           ...  ..  ...   
4805                 y  no_data        Indoor  40  ...   
4807                 n  no_data       no_data   2  ...   
4809                 y  no_data       Outdoor  28  ...   
4810                 y  no_data    Housing...  71  ...   
4811                 y  no_data    Housing...   9  ...   

                          LIBRARY_CONSTRUCTION_PROTOCOL      PKU  \
1                                               no_data       no   
4                                               no_data       no   
7                                               no_data       no   
9                                               no_data       no   
10    This analysis was done as in Caporaso et al 20...       no   
...                                                 ...      ...   
4805                                            no_data       no   
4807                                            no_data       no   
4809                                            no_data       no   
4810                                            no_data       no   
4811                                            no_data  unknown   

     DECEASED_PARENT  SEQ_RUN      CAT LATITUDE FOODALLERGIES_PEANUTS  \
1            no_data  no_data      yes     32.7               no_data   
4            no_data  no_data       no     43.1               no_data   
7            unknown  no_data      yes     44.7               unknown   
9            unknown  no_data  unknown     37.8               unknown   
10           unknown  no_data      yes       40               unknown   
...              ...      ...      ...      ...                   ...   
4805         unknown  no_data      yes     34.1               unknown   
4807         no_data  no_data      yes     40.3               no_data   
4809         unknown  no_data       no     39.3               unknown   
4810         unknown  no_data      yes       40               unknown   
4811         unkn

In [30]:
df[(df == 'unknown').any(axis=1)].shape

/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


(1557, 202)

In [31]:
# Look for rows, that contain cells labelled as 'Unknown'.

df[(df == 'Unknown').any(axis=1)]

#SampleID BarcodeSequence LinkerPrimerSequence CARBOHYDRATE_PER  \
2724  000012073.fixed49         no_data              no_data               60   

     NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS ASSIGNED_FROM_GEO  \
2724                   yes                    unknown                 y   

     RUN_DATE PET_LOCATIONS AGE  ... LIBRARY_CONSTRUCTION_PROTOCOL PKU  \
2724  no_data    Housing...  35  ...                       no_data  no   

     DECEASED_PARENT  SEQ_RUN CAT LATITUDE FOODALLERGIES_PEANUTS CHICKENPOX  \
2724         unknown  no_data  no     37.7               unknown        yes   

     SEASONAL_ALLERGIES Description  
2724                 no        None  

[1 rows x 202 columns]

In [32]:
# Look for rows, that contain cells labelled as 'None'.

df[(df == 'None').any(axis=1)]

/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


#SampleID BarcodeSequence LinkerPrimerSequence  \
10      000016722.fixed480    TTCCTGTTAACC  GTGCCAGCMGCCGCGGTAA   
12      000016449.fixed444    TGACAACCGAAT  GTGCCAGCMGCCGCGGTAA   
17      000018297.fixed577    GTAACCACCACC  GTGCCAGCMGCCGCGGTAA   
44      000016738.fixed493    CTGTAAAGGTTG  GTGCCAGCMGCCGCGGTAA   
57       000013636.fixed67         no_data              no_data   
...                    ...             ...                  ...   
4800    000016290.fixed430    GTACTGAAGATC  GTGCCAGCMGCCGCGGTAA   
4802  BLANK.2.10A.fixed595    AAGACGTAGCGG  GTGCCAGCMGCCGCGGTAA   
4805    000016360.fixed111         no_data              no_data   
4809    000015346.fixed136         no_data              no_data   
4810    000013296.fixed213         no_data              no_data   

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \
10                 52               unknown                    unknown   
12                 20               unknown                    unknown   
17                 20               unknown                    unknown   
44            unknown               unknown                    unknown   
57                 51                   yes                    unknown   
...               ...                   ...                        ...   
4800               30               unknown                    unknown   
4802          unknown               unknown                    unknown   
4805               19                   yes                    unknown   
4809               65               unknown                    unknown   
4810               50               unknown                    unknown   

     ASSIGNED_FROM_GEO RUN_DATE PET_LOCATIONS      AGE  ...  \
10                   y  8/21/14    Housing...       32  ...   
12                   y  9/18/14    Housing...       49  ...   
17                   y  9/18/14        Indoor       60  ...   
44                   y  8/21/14    Housing...       35  ...   
57                   y  no_data    Housing...       55  ...   
...                ...      ...           ...      ...  ...   
4800                 y  8/21/14    Housing...        1  ...   
4802                 y  8/21/14       unknown  unknown  ...   
4805                 y  no_data        Indoor       40  ...   
4809                 y  no_data       Outdoor       28  ...   
4810                 y  no_data    Housing...       71  ...   

                          LIBRARY_CONSTRUCTION_PROTOCOL      PKU  \
10    This analysis was done as in Caporaso et al 20...       no   
12    This analysis was done as in Caporaso et al 20...  unknown   
17    This analysis was done as in Caporaso et al 20...       no   
44    This analysis was done as in Caporaso et al 20...       no   
57                                              no_data       no   
...                                                 ...      ...   
4800  This analysis was done as in Caporaso et al 20...       no   
4802  This analysis was done as in Caporaso et al 20...  unknown   
4805                                            no_data       no   
4809                                            no_data       no   
4810                                            no_data       no   

     DECEASED_PARENT  SEQ_RUN      CAT LATITUDE FOODALLERGIES_PEANUTS  \
10           unknown  no_data      yes       40               unknown   
12           unknown  no_data      yes     64.8               unknown   
17           unknown  no_data  unknown     41.7               unknown   
44           unknown  no_data       no       40               unknown   
57           unknown  no_data       no       39               unknown   
...              ...      ...      ...      ...                   ...   
4800         unknown  no_data       no     52.4               unknown   
4802         unknown  no_data  unknown       40               unknown   
4805         unknown  no_data      yes     34.1               unknown   
4809         unknow

In [33]:
df[(df == 'None').any(axis=1)].shape

/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


(623, 202)

In [34]:
# Look for rows, that contain cells labelled as 'none'.

df[(df == 'none').any(axis=1)]

#SampleID BarcodeSequence LinkerPrimerSequence  \
209   000016470.fixed656         no_data              no_data   
224          11335.12105         no_data              no_data   
764   000016471.fixed968         no_data              no_data   
788          7106.107567         no_data              no_data   
985          1763.107581         no_data              no_data   
1114         7102.107567         no_data              no_data   
1275         3812.113184         no_data              no_data   
1364  000016334.fixed235         no_data              no_data   
1526         11190.12107         no_data              no_data   
1687         6668.113033         no_data              no_data   
2045         6673.105393         no_data              no_data   
2646         2665.113158         no_data              no_data   
3502          7103.10757         no_data              no_data   
3515         5016.107642         no_data              no_data   
3664         7105.107565         no_data              no_data   
3725         3891.113168         no_data              no_data   
4084         14126.12619         no_data              no_data   
4266         7104.107568         no_data              no_data   
4580         13079.12571         no_data              no_data   
4592         6805.113051         no_data              no_data   

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \
209                66               unknown                    unknown   
224                63               no_data                    no_data   
764                66               unknown                    unknown   
788                40               no_data                    no_data   
985                 5               no_data                    no_data   
1114               40               no_data                    no_data   
1275               47               no_data                    no_data   
1364               53               unknown                    unknown   
1526               53               no_data                    no_data   
1687               10               no_data                    no_data   
2045               77                   yes                    no_data   
2646               30               no_data                    no_data   
3502               40               no_data                    no_data   
3515               48               no_data                    no_data   
3664               40               no_data                    no_data   
3725               20               no_data                        yes   
4084               42               no_data                    no_data   
4266               40               no_data                    no_data   
4580          no_data               no_data                    no_data   
4592               35               no_data                    no_data   

     ASSIGNED_FROM_GEO RUN_DATE             PET_LOCATIONS AGE  ...  \
209                  y  no_data                Housing...  21  ...   
224                  n  no_data                Housing...  71  ...   
764                  y  no_data                Housing...  21  ...   
788                  n  no_data                   no_data  51  ...   
985                  n  no_data                   no_data  35  ...   
1114                 n  no_data                   no_data  51  ...   
1275                 n  no_data                   no_data   2  ...   
1364                 y  no_data                Housing...  48  ...   
1526                 n  no_data                Housing...  62  ...   
1687                 n  no_data                   no_data  69  ...   
2045                 n  no_data                   no_data  64  ...   
2646                 n  no_data                   no_data  70  ...   
3502                 n  no_data                   no_data  51  ...   
3515                 n  no_data                    Indoor   4  ...   
3664                 n  no_data             

In [35]:
df[(df == 'none').any(axis=1)].shape

/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


(20, 202)

---

Inspect all columns using **domain knowledge to decide which ones are relevant**, in order to select which columns to delete. 

In [36]:
# Get the column names.
   
list(df.columns.values) 

# https://www.geeksforgeeks.org/how-to-get-column-names-in-pandas-dataframe/

['#SampleID',
 'BarcodeSequence',
 'LinkerPrimerSequence',
 'CARBOHYDRATE_PER',
 'NONFOODALLERGIES_DRUG',
 'NONFOODALLERGIES_BEESTINGS',
 'ASSIGNED_FROM_GEO',
 'RUN_DATE',
 'PET_LOCATIONS',
 'AGE',
 'TOT_MASS',
 'GENERAL_MEDS',
 'SAMPLE_ID',
 'ORIG_SAMPLE_NAME',
 'QUINOLINE',
 'BODY_SITE',
 'MIGRAINE_FACTOR_2',
 'MIGRAINE_FACTOR_3',
 'TONSILS_REMOVED',
 'MIGRAINE_FACTOR_1',
 'ELEVATION',
 'FLOSSING_FREQUENCY',
 'PROTEIN_PER',
 'CSECTION',
 'PCR_PRIMERS',
 'COUNTRY_OF_BIRTH',
 'MACRONUTRIENT_PCT_TOTAL',
 'COSMETICS_FREQUENCY',
 'MASTERMIX_LOT',
 'HOST_SUBJECT_ID',
 'ENA-CHECKLIST',
 'GLUTEN',
 'SAMPLE_NAME',
 'TM1000_8_TOOL',
 'LONGITUDE',
 'DOG',
 'MIGRAINE_AGGRAVATION',
 'ORIG_NAME',
 'FRAT',
 'DIABETES_MEDICATION',
 'SAMPLE_PLATE',
 'PENICILLIN',
 'FOODALLERGIES_OTHER_TEXT',
 'KEY_SEQ',
 'BODY_PRODUCT',
 'REGION',
 'SAMPLE_TYPE',
 'MIGRAINE_FREQUENCY',
 'MIGRAINE_RELATIVES',
 'EXPERIMENT_TITLE',
 'PLANT_PER',
 'EXTRACTIONKIT_LOT',
 'SAMPLE',
 'MIGRAINE_PHOTOPHOBIA',
 'SITE_SAMPLED',


In [37]:
# Show the top and bottom 5 rows of all columns. 

with pd.option_context('display.max_columns', None):
    print (df)

                #SampleID BarcodeSequence LinkerPrimerSequence  \
0             11224.12109         no_data              no_data   
1             5034.107635         no_data              no_data   
2             11351.12107         no_data              no_data   
3             2013.113155         no_data              no_data   
4             4731.107641         no_data              no_data   
...                   ...             ...                  ...   
4810   000013296.fixed213         no_data              no_data   
4811  000007822.fixed1101         no_data              no_data   
4812          9074.121304         no_data              no_data   
4813          2418.128313         no_data              no_data   
4814          7179.126173         no_data              no_data   

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS  \
0                  55               no_data                    no_data   
1                  56               no_data                

**Columns to delete:** 

In [38]:
# Import csv with the names of the columns to be deleted in deletion round 1.

df_delete_1 = pd.read_csv("columns_delete_in_step_1.csv")
df_delete_1.head(5)

delete_in_step_1
0             #SampleID
1       BarcodeSequence
2  LinkerPrimerSequence
3     ASSIGNED_FROM_GEO
4              RUN_DATE

In [39]:
# Convert dataframe into a list. 

df_delete_1_list = df_delete_1['delete_in_step_1'].tolist()
df_delete_1_list

# https://www.geeksforgeeks.org/python-pandas-index-tolist/#:~:text=tolist()%20function%20return%20a,Timedelta%2FInterval%2FPeriod).

['#SampleID',
 'BarcodeSequence',
 'LinkerPrimerSequence',
 'ASSIGNED_FROM_GEO',
 'RUN_DATE',
 'PET_LOCATIONS',
 'GENERAL_MEDS',
 'SAMPLE_ID',
 'ORIG_SAMPLE_NAME',
 'PCR_PRIMERS',
 'MASTERMIX_LOT',
 'HOST_SUBJECT_ID',
 'ENA-CHECKLIST',
 'SAMPLE_NAME',
 'TM1000_8_TOOL',
 'ORIG_NAME',
 'FRAT',
 'SAMPLE_PLATE',
 'FOODALLERGIES_OTHER_TEXT',
 'KEY_SEQ',
 'BODY_PRODUCT',
 'REGION',
 'SAMPLE_TYPE',
 'EXPERIMENT_TITLE',
 'EXTRACTIONKIT_LOT',
 'SAMPLE',
 'CENTER_NAME',
 'COMMON_NAME',
 'SEQUENCING_METH',
 'ENV_BIOME',
 'WELL_ID',
 'SAMPLE_CENTER',
 'PHYSICAL_LOCATION',
 'LINKER',
 'TANNING_SPRAYS',
 'HOST_COMMON_NAME',
 'RACE_OTHER',
 'ALTITUDE',
 'NONFOODALLERGIES_DANDER',
 'TARGET_SUBFRAGMENT',
 'TITLE',
 'MIGRAINE_PAIN',
 'ROOMMATES',
 'CENTER_PROJECT_NAME',
 'HAS_EXTRACTED_DATA',
 'ENA-SPOT-COUNT',
 'RUN_ID',
 'COLLECTION_DATE',
 'TARGET_GENE',
 'PREGNANT_DUE_DATE',
 'SAMPLE_DESCRIPTION',
 'SAMP_SIZE',
 'HAS_PHYSICAL_SPECIMEN',
 'STUDY_CENTER',
 'WATER_LOT',
 'ENV_FEATURE',
 'ENA-BASE-COUNT

In [40]:
len(df_delete_1_list)

92

In [41]:
# Delete columns shown in 'df_delete_1_list'. 

df_2 = df.drop(df_delete_1_list, axis = 1) 

In [42]:
# Get shape of updated data frame df_2 after deleting 92 columns.

df_2.shape

(4815, 110)

df_2 has 4815 rows and 110 columns. This means that 45.5 % of the columns have been deleted at this step. 

In [43]:
# Checking the shape of the original data frame df.

df.shape

(4815, 202)

In [44]:
# Get the column names of df_2, the new smaller data frame.
   
df_2_list = list(df_2.columns.values) 
df_2_list

# https://www.geeksforgeeks.org/how-to-get-column-names-in-pandas-dataframe/

['CARBOHYDRATE_PER',
 'NONFOODALLERGIES_DRUG',
 'NONFOODALLERGIES_BEESTINGS',
 'AGE',
 'TOT_MASS',
 'QUINOLINE',
 'BODY_SITE',
 'MIGRAINE_FACTOR_2',
 'MIGRAINE_FACTOR_3',
 'TONSILS_REMOVED',
 'MIGRAINE_FACTOR_1',
 'ELEVATION',
 'FLOSSING_FREQUENCY',
 'PROTEIN_PER',
 'CSECTION',
 'COUNTRY_OF_BIRTH',
 'MACRONUTRIENT_PCT_TOTAL',
 'COSMETICS_FREQUENCY',
 'GLUTEN',
 'LONGITUDE',
 'DOG',
 'MIGRAINE_AGGRAVATION',
 'DIABETES_MEDICATION',
 'PENICILLIN',
 'MIGRAINE_FREQUENCY',
 'MIGRAINE_RELATIVES',
 'PLANT_PER',
 'MIGRAINE_PHOTOPHOBIA',
 'SITE_SAMPLED',
 'DRINKING_WATER_SOURCE',
 'COUNTRY',
 'ANTIBIOTIC_SELECT',
 'MIGRAINE_NAUSEA',
 'FOODALLERGIES_TREENUTS',
 'MIGRAINE_AURA',
 'FLU_VACCINE_DATE',
 'PREGNANT',
 'MIGRAINE',
 'BMI',
 'TEETHBRUSHING_FREQUENCY',
 'MIGRAINE_MEDICATIONS',
 'CONTRACEPTIVE',
 'LIVINGWITH',
 'RELATIONS',
 'DIABETES_MEDICATIONS',
 'SUPPLEMENTS',
 'AGE_UNIT',
 'EXERCISE_FREQUENCY',
 'SULFA_DRUG',
 'NONFOODALLERGIES_NO',
 'ENV_MATTER',
 'SKIN_CONDITION',
 'CONDITIONS_MEDICA

In [45]:
# Show the top and bottom 5 rows of all columns in df_2, the new smaller data frame.

with pd.option_context('display.max_columns', None):
    print (df_2)

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS AGE  \
0                  55               no_data                    no_data  51   
1                  56               no_data                    no_data  35   
2                  16               no_data                    no_data  56   
3             no_data               no_data                    no_data  37   
4                  61                   yes                    no_data  72   
...               ...                   ...                        ...  ..   
4810               50               unknown                    unknown  71   
4811               17               unknown                    unknown   9   
4812               34               no_data                    no_data  29   
4813               30               no_data                    no_data  46   
4814               60               no_data                    no_data   1   

     TOT_MASS QUINOLINE      BODY_SITE MIGRAINE_FACTOR_2 MIGRAI

In [46]:
# Check if there are any duplicate columns.

df_2.columns.value_counts()

# https://cmdlinetips.com/2020/08/pandas-value_counts-how-to-get-frequency-counts-of-variables-in-a-dataframe/

GLUTEN                 1
DIABETES_MEDICATION    1
ANIMAL_PER             1
SITE_SAMPLED           1
ACNE_MEDICATION        1
                      ..
TANNING_BEDS           1
NONFOODALLERGIES_NO    1
SEASONAL_ALLERGIES     1
TOT_MASS               1
FLOSSING_FREQUENCY     1
Length: 110, dtype: int64

Length is 110, i.e each column name appears only once, hence there are no duplicate columns.

---

The dataset contains several rows that correspond to blank samples and not to  a person. 
These rows are removed below, creating df_3.

In [47]:
# Remove rows based on two conditions.

df_3 = df_2.drop(df_2[(df_2['AGE'] == 'unknown') | (df_2['AGE'] == 'no_data')].index)
df_3.head(40)

# https://stackoverflow.com/questions/13851535/how-to-delete-rows-from-a-pandas-dataframe-based-on-a-conditional-expression

CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS AGE  \
0                55               no_data                    no_data  51   
1                56               no_data                    no_data  35   
2                16               no_data                    no_data  56   
3           no_data               no_data                    no_data  37   
4                61                   yes                    no_data  72   
5                55               no_data                    no_data  65   
6                60               no_data                    no_data  34   
7                37                   yes                    unknown  65   
8                32                   yes                    no_data  72   
9                50                   yes                    unknown  54   
10               52               unknown                    unknown  32   
11               50               no_data                    no_data  22   
12               20               unknown                    unknown  49   
13               40                   yes                    no_data  66   
14               63               no_data                    no_data  26   
15               30               no_data                    no_data  35   
16               51               no_data                    no_data  50   
17               20               unknown                    unknown  60   
18               35               no_data                    no_data  40   
19          no_data                   yes                    no_data  63   
21               61               no_data                    no_data  51   
22               46               no_data                    no_data  32   
23               62               no_data                    no_data  50   
24               52                   yes                    no_data  64   
25               60               no_data                    no_data  67   
26               35                   yes                        yes  43   
27               45               no_data                    no_data  62   
28               12               no_data                    no_data  53   
29               48               no_data                    no_data  39   
30          no_data               no_data                    no_data   4   
31               15               no_data                    no_data  42   
32               25               no_data                    no_data  43   
33               50               no_data                    no_data  34   
34               49               no_data                    no_data  23   
36               58               unknown                    unknown  64   
37               25               no_data                    no_data  48   
39               35               no_data                    no_data  58   
40             48.5               no_data                    no_data  59   
41               59                   yes                    no_data  64   
42               48               no_data                    no_data  46   

   TOT_MASS QUINOLINE       BODY_SITE MIGRAINE_FACTOR_2 MIGRAINE_FACTOR_3  \
0        50        no    UBERON:feces           no_data           no_data   
1        63   no_data    UBERON:feces           no_data           no_data   
2        64        no    UBERON:feces           no_data           no_data   
3        90        no    UBERON:feces           no_data           no_data   
4        86   no_data    UBERON:feces           no_data           no_data   
5        83        no    UBERON:feces           no_data           no_data   
6       101        no    UBERON:feces        Depression          Caffeine   
7        73        no    UBERON:feces           unknown           unknown   
8        58        no   UBERON:tongue           no_data           no_data   
9        80        no    UBERON:feces            Stress           unknown   
10       54        no    UBERON:feces  

In [48]:
df_3.shape

(4241, 110)

In [49]:
df_2.shape

(4815, 110)

In [50]:
deleted_rows_blank = df_2.shape[0] - df_3.shape[0]

deleted_rows_blank

574

574 rows where deleted since they were just blank. The updated data frame df_3 has 4241 rows and 110 columns.

---

In [51]:
df_3_columns = list(df_3.columns.values) 

for c in df_3_columns: 
    print(c)

CARBOHYDRATE_PER
NONFOODALLERGIES_DRUG
NONFOODALLERGIES_BEESTINGS
AGE
TOT_MASS
QUINOLINE
BODY_SITE
MIGRAINE_FACTOR_2
MIGRAINE_FACTOR_3
TONSILS_REMOVED
MIGRAINE_FACTOR_1
ELEVATION
FLOSSING_FREQUENCY
PROTEIN_PER
CSECTION
COUNTRY_OF_BIRTH
MACRONUTRIENT_PCT_TOTAL
COSMETICS_FREQUENCY
GLUTEN
LONGITUDE
DOG
MIGRAINE_AGGRAVATION
DIABETES_MEDICATION
PENICILLIN
MIGRAINE_FREQUENCY
MIGRAINE_RELATIVES
PLANT_PER
MIGRAINE_PHOTOPHOBIA
SITE_SAMPLED
DRINKING_WATER_SOURCE
COUNTRY
ANTIBIOTIC_SELECT
MIGRAINE_NAUSEA
FOODALLERGIES_TREENUTS
MIGRAINE_AURA
FLU_VACCINE_DATE
PREGNANT
MIGRAINE
BMI
TEETHBRUSHING_FREQUENCY
MIGRAINE_MEDICATIONS
CONTRACEPTIVE
LIVINGWITH
RELATIONS
DIABETES_MEDICATIONS
SUPPLEMENTS
AGE_UNIT
EXERCISE_FREQUENCY
SULFA_DRUG
NONFOODALLERGIES_NO
ENV_MATTER
SKIN_CONDITION
CONDITIONS_MEDICATION
LAST_TRAVEL
HEIGHT_IN
ANTIBIOTIC_CONDITION
PET_CONTACT
NONFOODALLERGIES_SUN
EXERCISE_LOCATION
NITROIMIDAZOLE
DEODORANT_USE
FOODALLERGIES_SHELLFISH
SEX
PERCENTAGE_FROM_CARBS
FAT_PER
SHARED_HOUSING
SOFTENER


In [52]:
abs_no_data_in_col = df_3.CARBOHYDRATE_PER.value_counts()

percentage_no_data_in_col = round(df_3.CARBOHYDRATE_PER.value_counts(normalize=True), 2)

print(f'absolute number of unique entries in column: {abs_no_data_in_col}')

print(f'percentage of unique entries in column: {percentage_no_data_in_col}')

# https://cmdlinetips.com/2020/08/pandas-value_counts-how-to-get-frequency-counts-of-variables-in-a-dataframe/
# https://cmdlinetips.com/2020/08/pandas-value_counts-how-to-get-frequency-counts-of-variables-in-a-dataframe/
# https://www.w3schools.com/python/ref_func_round.asp
# https://realpython.com/python-f-strings/

absolute number of unique entries in column: no_data    498
50         284
40         210
30         207
unknown    192
          ... 
42.3         1
22.2         1
51.7         1
49.8         1
47.1         1
Name: CARBOHYDRATE_PER, Length: 157, dtype: int64
percentage of unique entries in column: no_data    0.12
50         0.07
40         0.05
30         0.05
unknown    0.05
           ... 
42.3       0.00
22.2       0.00
51.7       0.00
49.8       0.00
47.1       0.00
Name: CARBOHYDRATE_PER, Length: 157, dtype: float64


Keep this column.

In [53]:
abs_no_data_in_col = df_3.NONFOODALLERGIES_DRUG.value_counts()

percentage_no_data_in_col = round(df_3.NONFOODALLERGIES_DRUG.value_counts(normalize=True), 2)

print(f'absolute number of unique entries in column: {abs_no_data_in_col}')

print(f'percentage of unique entries in column: {percentage_no_data_in_col}')

absolute number of unique entries in column: no_data    2722
yes         792
unknown     727
Name: NONFOODALLERGIES_DRUG, dtype: int64
percentage of unique entries in column: no_data    0.64
yes        0.19
unknown    0.17
Name: NONFOODALLERGIES_DRUG, dtype: float64


Keep this column.

---

In [54]:
# Ask if there are any Nan in the data frame.

df_3.isnull().values.any()

# https://datatofish.com/check-nan-pandas-dataframe/

True

There are NaN values in the data frame.

In [55]:
# Ask how many Nan there are in the data frame.

df_3.isnull().sum().sum()

# https://datatofish.com/check-nan-pandas-dataframe/

9

There are 9 rows that contain NaN values.

In [56]:
# Show rows with NaN.

is_NaN = df_3.isnull()
row_has_NaN = is_NaN.any(axis=1)
rows_with_NaN = df_3[row_has_NaN]

print(rows_with_NaN)

# https://www.kite.com/python/answers/how-to-find-rows-with-nan-values-in-a-pandas-dataframe-in-python

     CARBOHYDRATE_PER NONFOODALLERGIES_DRUG NONFOODALLERGIES_BEESTINGS AGE  \
190                69               no_data                    no_data  14   
208                20               unknown                    unknown  37   
522                20                   yes                    unknown  39   
1369               40               unknown                    unknown  37   
1980          no_data               no_data                    no_data  45   
2500          no_data               no_data                    no_data  34   
3051          no_data               no_data                    no_data  34   
3306               20               unknown                    unknown  38   
3997               60               no_data                    no_data   1   

     TOT_MASS QUINOLINE     BODY_SITE MIGRAINE_FACTOR_2 MIGRAINE_FACTOR_3  \
190        72        no  UBERON:feces           no_data           no_data   
208        84        no  UBERON:feces           unknown          

Above are the 9 rows that contain NaN values.

In [57]:
# Show columns with NaN.

df_3.loc[:, df.isna().any()]

# https://stackoverflow.com/questions/36226083/how-to-find-which-columns-contain-any-nan-value-in-pandas-dataframe

ANTIBIOTIC_CONDITION    STATE                PRIMARY_VEGETABLE  \
0                 no_data       CO                          spinach   
1                 no_data       CA  tomatoes (I eat a lot of salsa)   
2                 no_data       CO                     leafy greens   
3                 no_data       NY                           Potato   
4         dental cleaning       NY                         broccoli   
...                   ...      ...                              ...   
4810                  UTI       IL                  Romaine Lettuce   
4811      Streptococcus A  unknown              tomatoes, cuncumber   
4812              no_data       WA                           potato   
4813              no_data       IL                     Leafy Greens   
4814              no_data   BANKOK                Leaves and fruits   

                                      DIET_RESTRICTIONS  
0                                               no_data  
1                                               no_data  
2                                            Grain free  
3                                               no_data  
4                                               no_data  
...                                                 ...  
4810                                            unknown  
4811                                            unknown  
4812  Ray Peat diet. no grains, beans, minimal veget...  
4813                                            no_data  
4814                                            no_data  

[4241 rows x 4 columns]

Above are the 4 columns that contain NaN values.

In [58]:
# Drop rows with NaN values. This results in df_4.

df_4 = df_3.dropna(how='any') 

In [59]:
df_4.shape

(4232, 110)

In [60]:
df_3.shape

(4241, 110)

In [61]:
deleted_rows_NaN = df_3.shape[0] - df_4.shape[0]

deleted_rows_NaN

9

9 rows where deleted since they contained NaN values. The updated data frame **df_4 has 4232 rows and 110 columns**.

In [62]:
# Ask if there are any Nan in the data frame.

df_4.isnull().values.any()

False

There are no more NaN values in the data frame df_4.

---

In [63]:
# Show the data types of all columns.

df_4.dtypes

CARBOHYDRATE_PER              object
NONFOODALLERGIES_DRUG         object
NONFOODALLERGIES_BEESTINGS    object
AGE                           object
TOT_MASS                      object
                               ...  
CAT                           object
LATITUDE                      object
FOODALLERGIES_PEANUTS         object
CHICKENPOX                    object
SEASONAL_ALLERGIES            object
Length: 110, dtype: object

Keep data types as objects for first round of modelling.

---

# Changing of data type and feature engineering 

Change data type from object to float for 3 columns initially. Later on this will be done with more columns. 

**AGE**

In [64]:
df_4['AGE'].dtypes

dtype('O')

In [65]:
df_5 = df_4.astype({'AGE': 'float64'})

# https://thispointer.com/pandas-change-data-type-of-single-or-multiple-columns-of-dataframe-in-python/

In [66]:
df_5['AGE'].dtypes

dtype('float64')

---

**LATITUDE** - **IMPORTANT** for **Vitamin D3** levels

In [67]:
# Check data type

df_5['LATITUDE'].dtypes

dtype('O')

In [68]:
# Trying to change data type does not work: "ValueError: could not convert string to float: 'no_data'."

# df_6 = df_5.astype({'LATITUDE': 'float64'})

ValueError: could not convert string to float: 'no_data'

In [70]:
# Check data type is unchanged (object).

df_5['LATITUDE'].dtypes

dtype('O')

In [71]:
abs_no_data_in_col = (df_2.LATITUDE == 'no_data').sum()

percentage_no_data_in_col = round(((((df_2.LATITUDE == 'no_data').sum())/df_2.shape[0])*100), 3)

print(f'absolute number of no_data in column: {abs_no_data_in_col}')
print(f'percentage of no_data in column: {percentage_no_data_in_col}')

# https://stackoverflow.com/questions/35277075/python-pandas-counting-the-occurrences-of-a-specific-value

absolute number of no_data in column: 6
percentage of no_data in column: 0.125


In [72]:
df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]


LATITUDE COUNTRY_OF_BIRTH                       COUNTRY    STATE  \
217   no_data          no_data  GAZ:United States of America       MA   
678   no_data              USA  GAZ:United States of America       MA   
863   no_data      Switzerland               GAZ:Switzerland  no_data   
1531  no_data              USA  GAZ:United States of America       MA   
3325  no_data              USA  GAZ:United States of America       NY   
3972  no_data      Switzerland               GAZ:Switzerland  no_data   

     CURRENT_RESIDENCE_DURATION  
217      More than one year ago  
678      More than one year ago  
863      More than one year ago  
1531     More than one year ago  
3325     More than one year ago  
3972     More than one year ago

In [73]:
# Show unique entries.

df_5.STATE.unique()

# https://chrisalbon.com/python/data_wrangling/pandas_list_unique_values_in_column/

array(['CO', 'CA', 'NY', 'MA', 'MN', 'VERMONT', 'FL', 'AK', 'PA', 'NC',
       'MASSACHUSETTS', 'WI', 'NJ', 'OH', 'KY', 'TX', 'OR', '-', 'AZ',
       'MO', 'BANKOK', 'MD', 'GA', 'CT', 'VA', 'IN', 'NH', 'ME',
       'unknown', 'SC', 'IL', 'KS', 'NM', 'UT', 'WA', 'ILLINOIS',
       'LONDON', 'MB', 'ND', 'HI', 'BRITISH COLUMBIA', 'LA', 'OK',
       'TEXAS', 'BC', 'ALASKA', 'CALIFORNIA', 'IA', 'no_data', 'TN',
       'WASHINGTON', 'CEREDIGION', 'GREATER LONDON', 'CAMDEN TOWN', 'NV',
       'DC', 'MI', 'COLORADO', 'WV', 'AL', 'AR', 'ID', 'VIRGINIA', 'AB',
       'TENNESSEE', 'VT', 'MARYLAND', 'SD', 'NE', 'MT', 'MS', 'NEW YORK',
       'LONDON, CITY OF', 'WY', 'Bangkok', 'OHIO', 'CA - CALIFORNIA',
       'QC', 'RI', 'ONTARIO', 'NSW', 'ROTHERHITHE', 'NONE', 'MAINE',
       'OREGON', 'WISCONSIN', 'CONNECTICUT', 'SK', 'BUCKINGHAMSHIRE',
       'WESTERN AUSTRALIA', 'FLORIDA', 'HAWAII', 'GA.', 'NEW MEXICO',
       'VIC', 'MA - MASSACHUSETTS', 'COLORADO (CO)', 'SOUTH CAROLINA',
       'KENTUCKY', 

In [74]:
df_state_MA = df_5[df_5['STATE'] == 'MA'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]
df_state_MA

LATITUDE COUNTRY_OF_BIRTH                       COUNTRY STATE  \
5        42.8              USA  GAZ:United States of America    MA   
43       41.9              USA  GAZ:United States of America    MA   
61       42.3              USA  GAZ:United States of America    MA   
79       41.6              USA  GAZ:United States of America    MA   
105      42.3              USA  GAZ:United States of America    MA   
...       ...              ...                           ...   ...   
4665     42.4              USA  GAZ:United States of America    MA   
4703     42.3              USA  GAZ:United States of America    MA   
4724     42.4   United Kingdom  GAZ:United States of America    MA   
4778     41.6              USA  GAZ:United States of America    MA   
4790     42.6              USA  GAZ:United States of America    MA   

     CURRENT_RESIDENCE_DURATION  
5        More than one year ago  
43       More than one year ago  
61       More than one year ago  
79       More than one year ago  
105      More than one year ago  
...                         ...  
4665     More than one year ago  
4703     More than one year ago  
4724     More than one year ago  
4778     More than one year ago  
4790     More than one year ago  

[197 rows x 5 columns]

In [75]:
df_state_MA = df_5[df_5['STATE'] == 'MA'][['LATITUDE', 'STATE']]
df_state_MA.head(8)

LATITUDE STATE
5       42.8    MA
43      41.9    MA
61      42.3    MA
79      41.6    MA
105     42.3    MA
112     40.8    MA
178     41.6    MA
188     42.4    MA

In [76]:
df_state_MA_8 = df_state_MA.head(8)

In [77]:
df_state_MA_8['LATITUDE'].dtypes

dtype('O')

In [78]:
df_state_MA_8_float = df_state_MA_8.astype({'LATITUDE': 'float64'})

In [79]:
df_state_MA_8_float['LATITUDE'].dtypes

dtype('float64')

In [80]:
df_state_MA_8_float.describe()

LATITUDE
count    8.0000
mean    41.9625
std      0.6255
min     40.8000
25%     41.6000
50%     42.1000
75%     42.3250
max     42.8000

In [81]:
mean_latitude_MA = round(df_state_MA_8_float.mean(axis=0), 1)
mean_latitude_MA

# https://stackoverflow.com/questions/31037298/pandas-get-column-average-mean

LATITUDE    42.0
dtype: float64

Use this calculated mean value of LATITUDE = 42.0 for the cells that have STATE = MA and have 'no_data' under LATITUDE.

In [82]:
# Replace value in a cell.

df_5_1 = df_5.loc[217, 'LATITUDE'] = 42.0

df_5_1

42.0

In [83]:
df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]


LATITUDE COUNTRY_OF_BIRTH                       COUNTRY    STATE  \
678   no_data              USA  GAZ:United States of America       MA   
863   no_data      Switzerland               GAZ:Switzerland  no_data   
1531  no_data              USA  GAZ:United States of America       MA   
3325  no_data              USA  GAZ:United States of America       NY   
3972  no_data      Switzerland               GAZ:Switzerland  no_data   

     CURRENT_RESIDENCE_DURATION  
678      More than one year ago  
863      More than one year ago  
1531     More than one year ago  
3325     More than one year ago  
3972     More than one year ago

Replacement worked for that cell.

In [84]:
# Replace value in a cell.

df_5_2 = df_5.loc[678, 'LATITUDE'] = 42.0

df_5_3 = df_5.loc[1531, 'LATITUDE'] = 42.0

In [85]:
df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]


LATITUDE COUNTRY_OF_BIRTH                       COUNTRY    STATE  \
863   no_data      Switzerland               GAZ:Switzerland  no_data   
3325  no_data              USA  GAZ:United States of America       NY   
3972  no_data      Switzerland               GAZ:Switzerland  no_data   

     CURRENT_RESIDENCE_DURATION  
863      More than one year ago  
3325     More than one year ago  
3972     More than one year ago

Replacement worked for those 2 cells.

---

In [86]:
df_state_NY = df_5[df_5['STATE'] == 'NY'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]
df_state_NY

LATITUDE COUNTRY_OF_BIRTH                       COUNTRY STATE  \
3        40.7              USA  GAZ:United States of America    NY   
4        43.1              USA  GAZ:United States of America    NY   
19       40.7              USA  GAZ:United States of America    NY   
21         43              USA  GAZ:United States of America    NY   
25       40.7              USA  GAZ:United States of America    NY   
...       ...              ...                           ...   ...   
4761     42.6              USA  GAZ:United States of America    NY   
4762     44.3           Greece  GAZ:United States of America    NY   
4776     40.7              USA  GAZ:United States of America    NY   
4792     40.7              USA  GAZ:United States of America    NY   
4798     42.6              USA  GAZ:United States of America    NY   

     CURRENT_RESIDENCE_DURATION  
3        More than one year ago  
4        More than one year ago  
19       More than one year ago  
21       More than one year ago  
25       More than one year ago  
...                         ...  
4761     More than one year ago  
4762     More than one year ago  
4776     More than one year ago  
4792     More than one year ago  
4798     More than one year ago  

[240 rows x 5 columns]

In [87]:
df_state_NY = df_5[df_5['STATE'] == 'NY'][['LATITUDE', 'STATE']]
df_state_NY.head(30)

LATITUDE STATE
3       40.7    NY
4       43.1    NY
19      40.7    NY
21        43    NY
25      40.7    NY
54      40.8    NY
55      40.7    NY
160     40.7    NY
201     42.2    NY
210     40.7    NY
232     40.8    NY
238     40.9    NY
243     40.7    NY
254     40.9    NY
288     43.1    NY
290     40.7    NY
294     40.8    NY
310     40.7    NY
339     40.7    NY
345     42.7    NY
352     40.8    NY
378     40.7    NY
431     41.1    NY
483     40.7    NY
566     43.1    NY
567     40.7    NY
587     41.3    NY
600       43    NY
608     40.8    NY
618     40.8    NY

In [88]:
df_state_NY_30 = df_state_NY.head(30)

In [89]:
df_state_NY_30['LATITUDE'].dtypes

dtype('O')

In [90]:
df_state_NY_30_float = df_state_NY_30.astype({'LATITUDE': 'float64'})

In [91]:
df_state_NY_30_float['LATITUDE'].dtypes

dtype('float64')

In [92]:
df_state_NY_30_float.describe()

LATITUDE
count  30.000000
mean   41.276667
std     0.924637
min    40.700000
25%    40.700000
50%    40.800000
75%    41.250000
max    43.100000

In [93]:
mean_latitude_NY = round(df_state_NY_30_float.mean(axis=0), 1)
mean_latitude_NY

# https://stackoverflow.com/questions/31037298/pandas-get-column-average-mean

LATITUDE    41.3
dtype: float64

Use this calculated mean value of LATITUDE = 41.3 for the cells that have STATE = NY and have 'no_data' under LATITUDE.

In [94]:
# Replace value in a cell.

df_5_4 = df_5.loc[3325, 'LATITUDE'] = 41.3

df_5_4

41.3

In [95]:
df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]


LATITUDE COUNTRY_OF_BIRTH          COUNTRY    STATE  \
863   no_data      Switzerland  GAZ:Switzerland  no_data   
3972  no_data      Switzerland  GAZ:Switzerland  no_data   

     CURRENT_RESIDENCE_DURATION  
863      More than one year ago  
3972     More than one year ago

Replacement worked for that cell.

---

In [96]:
# Check if Switzerland is mentioned in 'STATE' column.

df_5[df_5['STATE'] == 'Switzerland']

Empty DataFrame
Columns: [CARBOHYDRATE_PER, NONFOODALLERGIES_DRUG, NONFOODALLERGIES_BEESTINGS, AGE, TOT_MASS, QUINOLINE, BODY_SITE, MIGRAINE_FACTOR_2, MIGRAINE_FACTOR_3, TONSILS_REMOVED, MIGRAINE_FACTOR_1, ELEVATION, FLOSSING_FREQUENCY, PROTEIN_PER, CSECTION, COUNTRY_OF_BIRTH, MACRONUTRIENT_PCT_TOTAL, COSMETICS_FREQUENCY, GLUTEN, LONGITUDE, DOG, MIGRAINE_AGGRAVATION, DIABETES_MEDICATION, PENICILLIN, MIGRAINE_FREQUENCY, MIGRAINE_RELATIVES, PLANT_PER, MIGRAINE_PHOTOPHOBIA, SITE_SAMPLED, DRINKING_WATER_SOURCE, COUNTRY, ANTIBIOTIC_SELECT, MIGRAINE_NAUSEA, FOODALLERGIES_TREENUTS, MIGRAINE_AURA, FLU_VACCINE_DATE, PREGNANT, MIGRAINE, BMI, TEETHBRUSHING_FREQUENCY, MIGRAINE_MEDICATIONS, CONTRACEPTIVE, LIVINGWITH, RELATIONS, DIABETES_MEDICATIONS, SUPPLEMENTS, AGE_UNIT, EXERCISE_FREQUENCY, SULFA_DRUG, NONFOODALLERGIES_NO, ENV_MATTER, SKIN_CONDITION, CONDITIONS_MEDICATION, LAST_TRAVEL, HEIGHT_IN, ANTIBIOTIC_CONDITION, PET_CONTACT, NONFOODALLERGIES_SUN, EXERCISE_LOCATION, NITROIMIDAZOLE, DEODORANT_USE, FOODALLERGIES_SHELLFISH, SEX, PERCENTAGE_FROM_CARBS, FAT_PER, SHARED_HOUSING, SOFTENER, MIGRAINE_PHONOPHOBIA, SMOKING_FREQUENCY, WEIGHT_LBS, TANNING_BEDS, ALCOHOL_FREQUENCY, STATE, BODY_HABITAT, DOMINANT_HAND, PRIMARY_VEGETABLE, SPECIAL_RESTRICTIONS, CEPHALOSPORIN, SUPPLEMENTS_FIELDS, APPENDIX_REMOVED, CURRENT_RESIDENCE_DURATION, SLEEP_DURATION, DIET_TYPE, DIET_RESTRICTIONS, LACTOSE, HEIGHT_OR_LENGTH, ANTIBIOTIC_MEDS, IBD, TYPES_OF_PLANTS, MIGRAINEMEDS, FIBER_GRAMS, NAILS, RACE, PETS, MULTIVITAMIN, NONFOODALLERGIES_POISONIVY, ACNE_MEDICATION, ACNE_MEDICATION_OTC, ANIMAL_PER, FOODALLERGIES_OTHER, ...]
Index: []

[0 rows x 110 columns]

In [97]:
# Show frequency ranking of unique entries.

abs_in_col = df_5.STATE.value_counts()

percentage_in_col = round(df_5.STATE.value_counts(normalize=True), 2)

print(f'absolute number of unique entries in column: {abs_in_col}')


print(f'percentage of unique entries in column: {percentage_in_col}')

absolute number of unique entries in column: CA                    711
CO                    324
NY                    240
MA                    197
WA                    175
                     ... 
BUSKERUD                1
MA - MASSACHUSETTS      1
ALASKA                  1
MINNESOTA               1
QUEBEC                  1
Name: STATE, Length: 118, dtype: int64
percentage of unique entries in column: CA                    0.17
CO                    0.08
NY                    0.06
MA                    0.05
WA                    0.04
                      ... 
BUSKERUD              0.00
MA - MASSACHUSETTS    0.00
ALASKA                0.00
MINNESOTA             0.00
QUEBEC                0.00
Name: STATE, Length: 118, dtype: float64


In [98]:
# Show unique entries.

df_5.STATE.unique()

array(['CO', 'CA', 'NY', 'MA', 'MN', 'VERMONT', 'FL', 'AK', 'PA', 'NC',
       'MASSACHUSETTS', 'WI', 'NJ', 'OH', 'KY', 'TX', 'OR', '-', 'AZ',
       'MO', 'BANKOK', 'MD', 'GA', 'CT', 'VA', 'IN', 'NH', 'ME',
       'unknown', 'SC', 'IL', 'KS', 'NM', 'UT', 'WA', 'ILLINOIS',
       'LONDON', 'MB', 'ND', 'HI', 'BRITISH COLUMBIA', 'LA', 'OK',
       'TEXAS', 'BC', 'ALASKA', 'CALIFORNIA', 'IA', 'no_data', 'TN',
       'WASHINGTON', 'CEREDIGION', 'GREATER LONDON', 'CAMDEN TOWN', 'NV',
       'DC', 'MI', 'COLORADO', 'WV', 'AL', 'AR', 'ID', 'VIRGINIA', 'AB',
       'TENNESSEE', 'VT', 'MARYLAND', 'SD', 'NE', 'MT', 'MS', 'NEW YORK',
       'LONDON, CITY OF', 'WY', 'Bangkok', 'OHIO', 'CA - CALIFORNIA',
       'QC', 'RI', 'ONTARIO', 'NSW', 'ROTHERHITHE', 'NONE', 'MAINE',
       'OREGON', 'WISCONSIN', 'CONNECTICUT', 'SK', 'BUCKINGHAMSHIRE',
       'WESTERN AUSTRALIA', 'FLORIDA', 'HAWAII', 'GA.', 'NEW MEXICO',
       'VIC', 'MA - MASSACHUSETTS', 'COLORADO (CO)', 'SOUTH CAROLINA',
       'KENTUCKY', 

In [100]:
# Enter ZURICH instead of no_data

df_state_ZURICH = df_5[df_5['STATE'] == 'ZURICH'][['LATITUDE', 'STATE']]
df_state_ZURICH.head(8)

LATITUDE   STATE
2603     47.7  ZURICH

In [101]:
# Replace value in a cell.

df_5_5 = df_5.loc[863, 'LATITUDE'] = 47.7

df_5_6 = df_5.loc[3972, 'LATITUDE'] = 47.7

print(df_5_5)

df_5_6

47.7


47.7

In [102]:
df_5[df_5['LATITUDE'] == 'no_data'][['LATITUDE', 'COUNTRY_OF_BIRTH', 'COUNTRY', 'STATE', 'CURRENT_RESIDENCE_DURATION']]


Empty DataFrame
Columns: [LATITUDE, COUNTRY_OF_BIRTH, COUNTRY, STATE, CURRENT_RESIDENCE_DURATION]
Index: []

Replacement worked for those 2 cells.

---

In [103]:
df_5['LATITUDE'].dtypes

dtype('O')

In [104]:
df_6 = df_5.astype({'LATITUDE': 'float64'})

In [105]:
df_6['LATITUDE'].dtypes

dtype('float64')

**df_6 will be used for modelling.**

---

# **Machine Learning Modelling**

In [108]:
df_clean = df_6 

In [117]:
df_clean.shape

(4232, 110)

Check the value range of a few specially important columns. Here I am choosing columns relevant for diabetes.

In [182]:
# Show unique values in a column.
# Age range.

AGE_unique = np.sort(df_6.AGE.unique())
AGE_unique

# https://stackoverflow.com/questions/32072076/find-the-unique-values-in-a-column-and-then-sort-them/32072252

array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.,
       13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
       26., 27., 28., 29., 30., 31., 32., 33., 34., 35., 36., 37., 38.,
       39., 40., 41., 42., 43., 44., 45., 46., 47., 48., 49., 50., 51.,
       52., 53., 54., 55., 56., 57., 58., 59., 60., 61., 62., 63., 64.,
       65., 66., 67., 68., 69., 70., 71., 72., 73., 74., 75., 76., 77.,
       78., 79., 80., 81., 82., 83., 84., 88., 91., 93.])

In [23]:
DIABETES_unique = np.sort(df.DIABETES.unique())
DIABETES_unique

array(['I do not have diabetes', 'Type I', 'Type II', 'no_data',
       'unknown'], dtype=object)

In [24]:
CARBOHYDRATE_PER_unique = np.sort(df.CARBOHYDRATE_PER.unique())
CARBOHYDRATE_PER_unique

array(['0', '1', '10', '11', '11.4', '12', '13', '13.5', '14', '15', '16',
       '17', '17.3', '17.6', '18', '19', '19.2', '19.9', '2', '2.9', '20',
       '20.1', '21', '21.3', '22', '22.2', '23', '24', '25', '25.4', '26',
       '26.8', '27', '27.8', '28', '28.2', '28.3', '29', '29.2', '3',
       '30', '31', '31.4', '31.5', '31.7', '32', '33', '34', '34.2', '35',
       '35.01', '35.6', '36', '36.2', '36.3', '36.6', '36.61', '37', '38',
       '38.1', '38.7', '38.81', '39', '39.1', '39.4', '39.6', '39.8', '4',
       '4.2', '40', '40.2', '41', '41.6', '41.8', '42', '42.1', '42.2',
       '42.3', '42.6', '43', '43.3', '43.4', '43.7', '44', '44.3', '44.5',
       '44.7', '45', '45.4', '45.5', '46', '46.4', '46.5', '46.7', '47',
       '47.1', '47.2', '47.5', '47.6', '47.7', '48', '48.3', '48.5',
       '48.7', '49', '49.6', '49.8', '5', '50', '50.2', '50.8', '51',
       '51.7', '52', '53', '53.5', '54', '55', '55.5', '56', '56.1', '57',
       '57.7', '58', '59', '59.7', '6', '60', 

In [25]:
FAT_PER_unique = np.sort(df.FAT_PER.unique())
FAT_PER_unique

array(['0', '10', '11', '11.7', '12', '13', '13.9', '14', '15', '16',
       '17', '17.7', '18', '19', '19.01', '2', '20', '21', '22', '22.1',
       '23', '24', '25', '25.3', '25.5', '26', '26.8', '26.9', '27', '28',
       '28.3', '28.6', '29', '29.1', '29.4', '30', '30.5', '30.7', '30.9',
       '31', '31.2', '31.4', '31.6', '31.9', '32', '32.4', '32.7', '32.8',
       '33', '33.7', '33.8', '33.9', '34', '34.2', '34.5', '34.6', '35',
       '35.1', '35.4', '35.6', '36', '36.1', '36.2', '36.3', '36.5',
       '36.6', '37', '37.4', '37.5', '37.9', '38', '38.1', '38.3', '38.5',
       '38.8', '39', '39.1', '39.3', '39.5', '39.9', '4', '40', '40.3',
       '40.5', '40.94', '41', '41.2', '41.3', '41.4', '41.7', '42',
       '42.4', '42.6', '43', '43.2', '43.3', '44', '44.1', '44.8', '44.9',
       '45', '45.12', '45.3', '46', '46.58', '46.7', '47', '47.1', '47.2',
       '47.9', '48', '48.7', '49', '5', '50', '51', '51.3', '52', '53',
       '53.5', '54', '54.7', '55', '55.3', '55.8', '5

In [181]:
LATITUDE_unique = np.sort(df_6.LATITUDE.unique())
LATITUDE_unique

array([-41.3, -37.9, -37.8, -37.7, -36.8, -33.9, -32. , -31.9, -31.8,
       -30.6, -30.5,  13.7,  18.4,  19.6,  20.8,  20.9,  21.3,  21.4,
        21.9,  22.1,  22.4,  24.7,  25.5,  25.7,  25.8,  25.9,  26.1,
        26.2,  26.3,  26.4,  26.7,  27. ,  27.1,  27.2,  27.5,  27.6,
        27.7,  27.9,  28. ,  28.1,  28.2,  28.3,  28.5,  28.6,  28.7,
        28.8,  29.5,  29.6,  29.7,  29.8,  29.9,  30. ,  30.1,  30.2,
        30.3,  30.4,  30.5,  30.6,  30.7,  31.1,  31.2,  31.3,  31.5,
        31.9,  32. ,  32.1,  32.2,  32.3,  32.4,  32.5,  32.6,  32.7,
        32.8,  32.9,  33. ,  33.1,  33.2,  33.3,  33.4,  33.5,  33.6,
        33.7,  33.8,  33.9,  34. ,  34.1,  34.2,  34.3,  34.4,  34.5,
        34.6,  34.7,  34.8,  34.9,  35. ,  35.1,  35.2,  35.3,  35.4,
        35.5,  35.6,  35.7,  35.8,  35.9,  36. ,  36.1,  36.2,  36.3,
        36.4,  36.5,  36.6,  36.7,  36.8,  36.9,  37. ,  37.1,  37.2,
        37.3,  37.4,  37.5,  37.6,  37.7,  37.8,  37.9,  38. ,  38.1,
        38.2,  38.3,

In [172]:
df_6.loc[(df_6.FAT_PER == 'unknown'),'FAT_PER']='33.3'

# https://kanoki.org/2019/07/17/pandas-how-to-replace-values-based-on-conditions/

In [173]:
df_6.loc[(df_6.CARBOHYDRATE_PER == 'no_data'),'CARBOHYDRATE_PER']='33.3'

In [174]:
CARBOHYDRATE_PER_unique = np.sort(df_6.CARBOHYDRATE_PER.unique())
CARBOHYDRATE_PER_unique

array(['0', '10', '11', '11.4', '12', '13', '13.5', '14', '15', '16',
       '17', '17.3', '17.6', '18', '19', '19.2', '19.9', '2', '2.9', '20',
       '20.1', '21', '21.3', '22', '22.2', '23', '24', '25', '25.4', '26',
       '26.8', '27', '27.8', '28', '28.2', '28.3', '29', '29.2', '3',
       '30', '31', '31.4', '31.5', '32', '33', '33.3', '34', '34.2', '35',
       '35.01', '35.6', '36', '36.2', '36.3', '36.6', '36.61', '37', '38',
       '38.1', '38.7', '38.81', '39', '39.1', '39.4', '39.6', '39.8', '4',
       '4.2', '40', '40.2', '41', '41.6', '41.8', '42', '42.1', '42.2',
       '42.3', '42.6', '43', '43.3', '43.4', '43.7', '44', '44.3', '44.5',
       '44.7', '45', '45.4', '45.5', '46', '46.4', '46.5', '46.7', '47',
       '47.1', '47.2', '47.5', '47.6', '47.7', '48', '48.3', '48.5',
       '48.7', '49', '49.6', '49.8', '5', '50', '50.2', '50.8', '51',
       '51.7', '52', '53', '53.5', '54', '55', '55.5', '56', '56.1', '57',
       '57.7', '58', '59', '59.7', '6', '60', '61',

In [179]:
df_6.loc[(df_6.FAT_PER == 'unknown'),'FAT_PER']='33.3'

In [176]:
df_6.loc[(df_6.FAT_PER == 'no_data'),'FAT_PER']='33.3'

In [180]:
FAT_PER_unique = np.sort(df_6.FAT_PER.unique())
FAT_PER_unique

array(['0', '10', '11', '11.7', '12', '13', '13.9', '14', '15', '16',
       '17', '17.7', '18', '19', '19.01', '20', '21', '22', '22.1', '23',
       '24', '25', '25.3', '25.5', '26', '26.8', '26.9', '27', '28',
       '28.3', '28.6', '29', '29.1', '29.4', '30', '30.5', '30.7', '30.9',
       '31', '31.2', '31.4', '31.6', '31.9', '32', '32.4', '32.7', '32.8',
       '33', '33.3', '33.7', '33.8', '33.9', '34', '34.2', '34.5', '34.6',
       '35', '35.1', '35.4', '35.6', '36', '36.1', '36.2', '36.3', '36.5',
       '36.6', '37', '37.4', '37.5', '37.9', '38', '38.1', '38.3', '38.5',
       '38.8', '39', '39.1', '39.3', '39.5', '39.9', '4', '40', '40.3',
       '40.5', '40.94', '41', '41.2', '41.3', '41.4', '41.7', '42',
       '42.4', '42.6', '43', '43.2', '43.3', '44', '44.1', '44.8', '44.9',
       '45', '45.12', '45.3', '46', '46.58', '46.7', '47', '47.1', '47.2',
       '47.9', '48', '48.7', '49', '5', '50', '51', '51.3', '52', '53',
       '53.5', '54', '54.7', '55', '55.3', '55.8',

---

In [184]:
df_6['CARBOHYDRATE_PER'].dtypes

dtype('O')

In [185]:
df_7 = df_6.astype({'CARBOHYDRATE_PER': 'float64'})

In [186]:
df_7['LATITUDE'].dtypes

dtype('float64')

---

In [187]:
df_6['FAT_PER'].dtypes

dtype('O')

In [188]:
df_7 = df_6.astype({'FAT_PER': 'float64'})

In [189]:
df_7['FAT_PER'].dtypes

dtype('float64')

**df_7 will be used for modelling.**

In [195]:
df_clean = df_7

## **Target: DIABETES**

In [109]:
# These are all features that can potentially be used.

#X = df_clean[['CARBOHYDRATE_PER',
 'NONFOODALLERGIES_DRUG',
 'NONFOODALLERGIES_BEESTINGS',
 'AGE',
 'TOT_MASS',
 'QUINOLINE',
 'BODY_SITE',
 'MIGRAINE_FACTOR_2',
 'MIGRAINE_FACTOR_3',
 'TONSILS_REMOVED',
 'MIGRAINE_FACTOR_1',
 'ELEVATION',
 'FLOSSING_FREQUENCY',
 'PROTEIN_PER',
 'CSECTION',
 'COUNTRY_OF_BIRTH',
 'MACRONUTRIENT_PCT_TOTAL',
 'COSMETICS_FREQUENCY',
 'GLUTEN',
 'LONGITUDE',
 'DOG',
 'MIGRAINE_AGGRAVATION',
 'DIABETES_MEDICATION',
 'PENICILLIN',
 'MIGRAINE_FREQUENCY',
 'MIGRAINE_RELATIVES',
 'PLANT_PER',
 'MIGRAINE_PHOTOPHOBIA',
 'SITE_SAMPLED',
 'DRINKING_WATER_SOURCE',
 'COUNTRY',
 'ANTIBIOTIC_SELECT',
 'MIGRAINE_NAUSEA',
 'FOODALLERGIES_TREENUTS',
 'MIGRAINE_AURA',
 'FLU_VACCINE_DATE',
 'PREGNANT',
 'MIGRAINE',
 'BMI',
 'TEETHBRUSHING_FREQUENCY',
 'MIGRAINE_MEDICATIONS',
 'CONTRACEPTIVE',
 'LIVINGWITH',
 'RELATIONS',
 'DIABETES_MEDICATIONS',
 'SUPPLEMENTS',
 'AGE_UNIT',
 'EXERCISE_FREQUENCY',
 'SULFA_DRUG',
 'NONFOODALLERGIES_NO',
 'ENV_MATTER',
 'SKIN_CONDITION',
 'CONDITIONS_MEDICATION',
 'LAST_TRAVEL',
 'HEIGHT_IN',
 'ANTIBIOTIC_CONDITION',
 'PET_CONTACT',
 'NONFOODALLERGIES_SUN',
 'EXERCISE_LOCATION',
 'NITROIMIDAZOLE',
 'DEODORANT_USE',
 'FOODALLERGIES_SHELLFISH',
 'SEX',
 'PERCENTAGE_FROM_CARBS',
 'FAT_PER',
 'SHARED_HOUSING',
 'SOFTENER',
 'MIGRAINE_PHONOPHOBIA',
 'SMOKING_FREQUENCY',
 'WEIGHT_LBS',
 'TANNING_BEDS',
 'ALCOHOL_FREQUENCY',
 'STATE',
 'BODY_HABITAT',
 'DOMINANT_HAND',
 'PRIMARY_VEGETABLE',
 'SPECIAL_RESTRICTIONS',
 'CEPHALOSPORIN',
 'SUPPLEMENTS_FIELDS',
 'APPENDIX_REMOVED',
 'CURRENT_RESIDENCE_DURATION',
 'SLEEP_DURATION',
 'DIET_TYPE',
 'DIET_RESTRICTIONS',
 'LACTOSE',
 'HEIGHT_OR_LENGTH',
 'ANTIBIOTIC_MEDS',
 'IBD',
 'TYPES_OF_PLANTS',
 'MIGRAINEMEDS',
 'FIBER_GRAMS',
 'NAILS',
 'RACE',
 'PETS',
 'MULTIVITAMIN',
 'NONFOODALLERGIES_POISONIVY',
 'ACNE_MEDICATION',
 'ACNE_MEDICATION_OTC',
 'ANIMAL_PER',
 'FOODALLERGIES_OTHER',
 'WEIGHT_CHANGE',
 'PRIMARY_CARB',
 'ASTHMA',
 'POOL_FREQUENCY',
 'CAT',
 'LATITUDE',
 'FOODALLERGIES_PEANUTS',
 'CHICKENPOX',
 'SEASONAL_ALLERGIES']] # independent variables or features

#y = df_clean['DIABETES'] # dependent variable or target

In [276]:
# store clean data in X and y.

X = df_clean[['AGE',
 'LATITUDE',            
 'CARBOHYDRATE_PER',
 'FAT_PER']] # independent variables or features

y = df_clean['DIABETES'] # dependent variable or target

In [277]:
X.head(5)

AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER
0  51.0      39.7               55     30.0
1  35.0      32.7               56     21.0
2  56.0      40.0               16     67.0
3  37.0      40.7             33.3     33.3
4  72.0      43.1               61     23.0

In [278]:
X.shape

(4232, 4)

In [279]:
y.head(5)

0    I do not have diabetes
1    I do not have diabetes
2    I do not have diabetes
3    I do not have diabetes
4    I do not have diabetes
Name: DIABETES, dtype: object

In [280]:
y.shape

(4232,)

---

In [281]:
# splitting our data into train and test sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=19, stratify = y)

In [282]:
# size of training data

X_train.shape

(2835, 4)

In [283]:
# size of test data

X_test.shape

(1397, 4)

In [284]:
X_train.head(5)

AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER
1958  43.0      43.0               40     30.0
906   32.0      29.6               60     20.0
986   56.0      38.0               25     35.0
1202  46.0      47.5               30     30.0
3327  32.0      39.1               59     21.0

In [285]:
y_train.head(5)

1958    I do not have diabetes
906     I do not have diabetes
986     I do not have diabetes
1202    I do not have diabetes
3327                    Type I
Name: DIABETES, dtype: object

In [286]:
X_test.head(5)

AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER
1649  43.0      33.9             50.2     36.1
2042  35.0      41.7               34     18.0
3534  21.0      40.7               56     27.0
1902  46.0      47.5               30     30.0
3531  45.0      30.2               75     15.0

In [287]:
y_test.head(5)

1649    I do not have diabetes
2042    I do not have diabetes
3534    I do not have diabetes
1902    I do not have diabetes
3531    I do not have diabetes
Name: DIABETES, dtype: object

## **Logistic Regression model**

In [210]:
from sklearn.linear_model import LogisticRegression

In [288]:
# instantiate a logistic regression
logreg_DIABETES = LogisticRegression()

# fit the model to train set only
logreg_DIABETES.fit(X_train, y_train)

# score the model
print(f"train accuracy: {logreg_DIABETES.score(X_train, y_train)}")
print(f"test accuracy: {logreg_DIABETES.score(X_test, y_test)}")

train accuracy: 0.9488536155202821
test accuracy: 0.9491768074445239


/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [289]:
# Show frequency of distinct values in a column.

df_7['DIABETES'].value_counts()

I do not have diabetes    4016
no_data                     89
Type II                     78
Type I                      34
unknown                     15
Name: DIABETES, dtype: int64

In [290]:
df_7['DIABETES'].shape

(4232,)

---

**Add Scaling**

In [291]:
# Import scalar.
from sklearn.preprocessing import StandardScaler 

# The scaler is FIT ON THE TRAINING DATA ONLY 

# 2. make a scaler
ss_scaler = StandardScaler()

# 3. fit the scaler
ss_scaler.fit(X_train)

# 4. transform the data
X_train_scaled = ss_scaler.transform(X_train)
X_test_scaled = ss_scaler.transform(X_test)

In [292]:
# instantiate a logistic regression
logreg_scaled_DIABETES = LogisticRegression()

# fit the model to train set only
logreg_scaled_DIABETES.fit(X_train_scaled, y_train)

# score the model
print(f"train accuracy: {logreg_scaled_DIABETES.score(X_train_scaled, y_train)}")
print(f"test accuracy: {logreg_scaled_DIABETES.score(X_test_scaled, y_test)}")

train accuracy: 0.9488536155202821
test accuracy: 0.9491768074445239


No improvement of accuracy scores.

**Adding (C=0.01) as a definition of a hyperparameter**

In [293]:
# instantiate a logistic regression
logreg_scaled_DIABETES = LogisticRegression(C=1)
# C=1 is default value.

# fit the model to train set only
logreg_scaled_DIABETES.fit(X_train_scaled, y_train)

# score the model
print(f"train accuracy: {logreg_scaled_DIABETES.score(X_train_scaled, y_train)}")
print(f"test accuracy: {logreg_scaled_DIABETES.score(X_test_scaled, y_test)}")

train accuracy: 0.9488536155202821
test accuracy: 0.9491768074445239


No improvement of accuracy scores.

In [294]:
# instantiate a logistic regression
logreg_scaled_DIABETES = LogisticRegression(C=0.001)

# fit the model to train set only
logreg_scaled_DIABETES.fit(X_train_scaled, y_train)

# score the model
print(f"train accuracy: {logreg_scaled_DIABETES.score(X_train_scaled, y_train)}")
print(f"test accuracy: {logreg_scaled_DIABETES.score(X_test_scaled, y_test)}")

train accuracy: 0.9488536155202821
test accuracy: 0.9491768074445239


No improvement of accuracy scores.

The accuracy score on the test set is 0.949.

In [295]:
# Coefficient: it returns an array with one spot for each feature.

coefficient = logreg_scaled_DIABETES.coef_[0]

print(f'Coefficients: {coefficient}')

Coefficients: [-0.03618958  0.00222874  0.01805508  0.0161957 ]


In [296]:
coefficient.shape

(4,)

In [297]:
list(X_train.columns.values) 

['AGE', 'LATITUDE', 'CARBOHYDRATE_PER', 'FAT_PER']

In [298]:
from sklearn.metrics import classification_report

predicted = logreg_scaled_DIABETES.predict(X_test_scaled)
report = classification_report(y_test, predicted)
print(report)

# https://machinelearningmastery.com/metrics-evaluate-machine-learning-algorithms-python/

                        precision    recall  f1-score   support

I do not have diabetes       0.95      1.00      0.97      1326
                Type I       0.00      0.00      0.00        11
               Type II       0.00      0.00      0.00        26
               no_data       0.00      0.00      0.00        29
               unknown       0.00      0.00      0.00         5

              accuracy                           0.95      1397
             macro avg       0.19      0.20      0.19      1397
          weighted avg       0.90      0.95      0.92      1397



/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


---

## **DecisionTreeClassifier**

In [233]:
from sklearn.tree import DecisionTreeClassifier

In [234]:
#Fit to the training data
DT_model_scaled_DIABETES = DecisionTreeClassifier()
DT_model_scaled_DIABETES.fit(X_train_scaled, y_train)

print(f"The TRAIN classification accuracy is:  {DT_model_scaled_DIABETES.score(X_train_scaled,y_train)}")
print(f"The TEST classification accuracy is:  {DT_model_scaled_DIABETES.score(X_test_scaled,y_test)}")

The TRAIN classification accuracy is:  0.9982363315696648
The TEST classification accuracy is:  0.9377236936292055


The accuracy score on the test set is 0.938.

---

## **Confusion matrix**

In [235]:
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay

In [237]:
# Get class predictions
y_pred = logreg_scaled_DIABETES.predict(X_train_scaled)

# Generate confusion matrix
cf_matrix = confusion_matrix(y_test, y_pred)
cf_matrix

# Plot
ConfusionMatrixDisplay(cf_matrix, display_labels=['0','1']).plot()
plt.show()

ValueError: Found input variables with inconsistent numbers of samples: [1397, 2835]

In [238]:
# Importing confusion matrix.

confusion = confusion_matrix(y_test, y_pred)
print('Confusion Matrix\n')
print(confusion)

# https://towardsdatascience.com/confusion-matrix-for-your-multi-class-machine-learning-model-ff9aa3bf7826

ValueError: Found input variables with inconsistent numbers of samples: [1397, 2835]

---

## IBD - Crohn's Disease and Ulcerative Colitis are the two main forms of Inflammatory Bowel Disease (IBD).



In [240]:
# store clean data in X and y.

X = df_clean[['AGE',
 'LATITUDE',            
 'CARBOHYDRATE_PER',
 'FAT_PER']] # independent variables or features

y = df_clean['IBD'] # dependent variable or target

In [241]:
X.head(5)

AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER
0  51.0      39.7               55     30.0
1  35.0      32.7               56     21.0
2  56.0      40.0               16     67.0
3  37.0      40.7             33.3     33.3
4  72.0      43.1               61     23.0

In [242]:
X.shape

(4232, 4)

In [243]:
y.head(5)

0    I do not have IBD
1    I do not have IBD
2    I do not have IBD
3    I do not have IBD
4    I do not have IBD
Name: IBD, dtype: object

In [244]:
y.shape

(4232,)

---

In [246]:
# splitting our data into train and test sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=19, stratify = y)

In [247]:
# size of training data

X_train.shape

(2835, 4)

In [248]:
# size of test data

X_test.shape

(1397, 4)

In [249]:
X_train.head(5)

AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER
3213  31.0      34.1               20     50.0
4163  37.0      37.4               43     39.0
2186  65.0      20.8               64     21.0
3879  60.0      32.8             33.3     33.3
2778  54.0      40.7               49     36.0

In [250]:
y_train.head(5)

3213    I do not have IBD
4163    I do not have IBD
2186    I do not have IBD
3879    I do not have IBD
2778    I do not have IBD
Name: IBD, dtype: object

In [251]:
X_test.head(5)

AGE  LATITUDE CARBOHYDRATE_PER  FAT_PER
4724  38.0      42.4               56     25.0
2549  20.0      41.8               52     34.0
2040  70.0      43.1               53     26.0
1577  35.0      40.0             19.9     62.2
432   41.0      37.6               17     53.0

In [252]:
y_test.head(5)

4724    I do not have IBD
2549    I do not have IBD
2040    I do not have IBD
1577    I do not have IBD
432     I do not have IBD
Name: IBD, dtype: object

## **Logistic Regression model**

In [253]:
from sklearn.linear_model import LogisticRegression

In [254]:
# instantiate a logistic regression
logreg_IBD = LogisticRegression()

# fit the model to train set only
logreg_IBD.fit(X_train, y_train)

# score the model
print(f"train accuracy: {logreg_IBD.score(X_train, y_train)}")
print(f"test accuracy: {logreg_IBD.score(X_test, y_test)}")

train accuracy: 0.9199294532627866
test accuracy: 0.9198282032927703


/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [255]:
# Show frequency of distinct values in a column.

df_7['IBD'].value_counts()

I do not have IBD     3893
no_data                127
Ulcerative colitis     116
Crohn's disease         69
unknown                 27
Name: IBD, dtype: int64

In [256]:
df_7['IBD'].shape

(4232,)

---

**Add Scaling**

In [257]:
# Import scalar.
from sklearn.preprocessing import StandardScaler 

# The scaler is FIT ON THE TRAINING DATA ONLY 

# 2. make a scaler
ss_scaler = StandardScaler()

# 3. fit the scaler
ss_scaler.fit(X_train)

# 4. transform the data
X_train_scaled = ss_scaler.transform(X_train)
X_test_scaled = ss_scaler.transform(X_test)

In [258]:
# instantiate a logistic regression
logreg_scaled_IBD = LogisticRegression()

# fit the model to train set only
logreg_scaled_IBD.fit(X_train_scaled, y_train)

# score the model
print(f"train accuracy: {logreg_scaled_IBD.score(X_train_scaled, y_train)}")
print(f"test accuracy: {logreg_scaled_IBD.score(X_test_scaled, y_test)}")

train accuracy: 0.9199294532627866
test accuracy: 0.9198282032927703


No improvement of accuracy scores.

**Adding (C=0.01) as a definition of a hyperparameter**

In [259]:
# instantiate a logistic regression
logreg_scaled_IBD = LogisticRegression(C=1)
# C=1 is default value.

# fit the model to train set only
logreg_scaled_IBD.fit(X_train_scaled, y_train)

# score the model
print(f"train accuracy: {logreg_scaled_IBD.score(X_train_scaled, y_train)}")
print(f"test accuracy: {logreg_scaled_IBD.score(X_test_scaled, y_test)}")

train accuracy: 0.9199294532627866
test accuracy: 0.9198282032927703


No improvement of accuracy scores.

In [260]:
# instantiate a logistic regression
logreg_scaled_IBD = LogisticRegression(C=0.001)

# fit the model to train set only
logreg_scaled_IBD.fit(X_train_scaled, y_train)

# score the model
print(f"train accuracy: {logreg_scaled_IBD.score(X_train_scaled, y_train)}")
print(f"test accuracy: {logreg_scaled_IBD.score(X_test_scaled, y_test)}")

train accuracy: 0.9199294532627866
test accuracy: 0.9198282032927703


No improvement of accuracy scores.

The accuracy score on the test set is 0.920.

In [261]:
# Coefficient: it returns an array with one spot for each feature.

coefficient = logreg_scaled_IBD.coef_[0]

print(f'Coefficients: {coefficient}')

Coefficients: [-0.00155367  0.00063138 -0.01024778  0.00301673]


In [262]:
coefficient.shape

(4,)

In [263]:
list(X_train.columns.values) 

['AGE', 'LATITUDE', 'CARBOHYDRATE_PER', 'FAT_PER']

In [268]:
# Show frequency of distinct values in a column.

df_7['IBD'].value_counts()

I do not have IBD     3893
no_data                127
Ulcerative colitis     116
Crohn's disease         69
unknown                 27
Name: IBD, dtype: int64

In [267]:
from sklearn.metrics import classification_report

predicted = logreg_scaled_IBD.predict(X_test_scaled)
report = classification_report(y_test, predicted)
print(report)

# https://machinelearningmastery.com/metrics-evaluate-machine-learning-algorithms-python/

                    precision    recall  f1-score   support

   Crohn's disease       0.00      0.00      0.00        23
 I do not have IBD       0.92      1.00      0.96      1285
Ulcerative colitis       0.00      0.00      0.00        38
           no_data       0.00      0.00      0.00        42
           unknown       0.00      0.00      0.00         9

          accuracy                           0.92      1397
         macro avg       0.18      0.20      0.19      1397
      weighted avg       0.85      0.92      0.88      1397



/Users/Cris/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


---

## **DecisionTreeClassifier**

In [233]:
from sklearn.tree import DecisionTreeClassifier

In [234]:
#Fit to the training data
DT_model_scaled_DIABETES = DecisionTreeClassifier()
DT_model_scaled_DIABETES.fit(X_train_scaled, y_train)

print(f"The TRAIN classification accuracy is:  {DT_model_scaled_DIABETES.score(X_train_scaled,y_train)}")
print(f"The TEST classification accuracy is:  {DT_model_scaled_DIABETES.score(X_test_scaled,y_test)}")

The TRAIN classification accuracy is:  0.9982363315696648
The TEST classification accuracy is:  0.9377236936292055


The accuracy score on the test set is 0.938.

---

In [299]:
df_clean.shape

(4232, 110)